File: [0212][Hard][Word_Search_II]/Word_Search_II.py
# https://leetcode.com/problems/word-search-ii/discuss/59804/27-lines-uses-complex-numbers
# leetcode time     cost : 444 ms
# leetcode memory   cost : 43.4 MB 
class Solution:
    #  def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
    def findWords(self, board, words):

        root = {}
        for word in words:
            node = root
            for c in word:
                node = node.setdefault(c, {})
            node[None] = True
        board = {i + 1j*j: c
                 for i, row in enumerate(board)
                 for j, c in enumerate(row)}

        found = []
        def search(node, z, word):
            if node.pop(None, None):
                found.append(word)
            c = board.get(z)
            if c in node:
                board[z] = None
                for k in range(4):
                    search(node[c], z + 1j**k, word + c)
                board[z] = c
        for z in board:
            search(root, z, '')

        return found

def main():
    board = [["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]] 
    words = ["oath","pea","eat","rain"]             #expect is ["oath","eat"]
    obj = Solution()
    result = obj.findWords(board, words)
    assert result == ["oath","eat"], ["hint: result is wrong"]
    print("return result is :",result)
    
if __name__ =='__main__':
    main() 

File: [0212][Hard][Word_Search_II]/Word_Search_II.java
//https://leetcode.com/problems/word-search-ii/discuss/59780/Java-15ms-Easiest-Solution-(100.00)
// leetcode time     cost : 9 ms
// leetcode memory   cost : 47.8 MB 

import java.util.List;
import java.util.ArrayList;
class TrieNode {
    TrieNode[] next = new TrieNode[26];
    String word;
}
class Solution {
    public List<String> findWords(char[][] board, String[] words) {
        List<String> res = new ArrayList<>();
        TrieNode root = buildTrie(words);
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                dfs (board, i, j, root, res);
            }
        }
        return res;
    }

    public void dfs(char[][] board, int i, int j, TrieNode p, List<String> res) {
        char c = board[i][j];
        if (c == '#' || p.next[c - 'a'] == null) return;
        p = p.next[c - 'a'];
        if (p.word != null) {   // found one
            res.add(p.word);
            p.word = null;     // de-duplicate
        }

        board[i][j] = '#';
        if (i > 0) dfs(board, i - 1, j ,p, res); 
        if (j > 0) dfs(board, i, j - 1, p, res);
        if (i < board.length - 1) dfs(board, i + 1, j, p, res); 
        if (j < board[0].length - 1) dfs(board, i, j + 1, p, res); 
        board[i][j] = c;
    }

    public TrieNode buildTrie(String[] words) {
        TrieNode root = new TrieNode();
        for (String w : words) {
            TrieNode p = root;
            for (char c : w.toCharArray()) {
                int i = c - 'a';
                if (p.next[i] == null) p.next[i] = new TrieNode();
                p = p.next[i];
        }
        p.word = w;
        }
        return root;
    }

}

public class Word_Search_II {
    public static void main(String args[]) {
        char[][] board = {{'o','a','a','n'},{'e','t','a','e'},{'i','h','k','r'},{'i','f','l','v'}}; 
        String[] words = {"oath","pea","eat","rain"};                     // #expect is ["oath","eat"]
        Solution Solution_obj = new Solution();
        List<String> result = Solution_obj.findWords(board, words);
        System.out.println("In java code,return value is :" + (result));
    }
}